runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -314,7 +314,10 @@ MM_VLHGCAccessBarrier::jniGetPrimitiveArrayCritical(J9VMThread* vmThread, jarray
 				GC_SlotObject objectSlot(env->getOmrVM(), &indexableObjectModel->getArrayoidPointer(arrayObject)[0]);
 				data = objectSlot.readReferenceFromSlot();
 			} else {
-				/* Possible to reach here if arraylet leaf has one leaf and no elements in it */
+				/* Possible to reach here if arraylet leaf has one leaf and no elements in it.
+				 * Even though there are no elements in it the caller expects a non NULL value
+				 * therefore we just return the address after the object header. */
+				data = (void *)GC_SlotObject::addToSlotAddress((fomrobject_t*)arrayObject, 1, env->compressObjectReferences());
```

Comment:
I think it should be &indexableObjectModel->getArrayoidPointer(arrayObject) instead of just arrayObject.
We go one (the only, empty) arroyid slot past the beginning of the arroyoid, to get to the end of the object.

Reply:
You can store arroyid pointer (or ObjectSlot based on it) into a local at the beginning of 'if double mapping' block, since it's being referenced a few times.